From 75864ba21b28172e469d0d6ae5c5905c37f691c1 Mon Sep 17 00:00:00 2001
From: Grygorii Strashko <grygorii.strashko@ti.com>
Date: Mon, 8 Feb 2021 18:28:34 +0200
Subject: [PATCH 26/35] cpts: latch: disable tmr irq while latch is on

Disable TMR IRQ while PPS latch is on and re-enable
it when PPS latch is off.
The CPTS HW_TS IRQ will happen on TMR overflow, so
adjustment can be done from CPTS IRQ.

Signed-off-by: Grygorii Strashko <grygorii.strashko@ti.com>
---
 drivers/net/ethernet/ti/cpts.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/net/ethernet/ti/cpts.c b/drivers/net/ethernet/ti/cpts.c
index 49eba5ba3efc..a34d401580d5 100644
--- a/drivers/net/ethernet/ti/cpts.c
+++ b/drivers/net/ethernet/ti/cpts.c
@@ -1923,6 +1923,11 @@ static inline void cpts_latch_pps_stop(struct cpts *cpts)
 	WRITE_TCLR(cpts->odt2, v);
 
 	cpts->pps_latch_state = INIT;
+
+	/* enable tmr irq for init-  no overflow/hwts from now */
+	writel_relaxed(OMAP_TIMER_INT_CAPTURE, cpts->odt2->irq_ena);
+	__omap_dm_timer_write(cpts->odt2, OMAP_TIMER_WAKEUP_EN_REG,
+			      OMAP_TIMER_INT_CAPTURE, 0);
 }
 
 static inline void cpts_latch_pps_start(struct cpts *cpts)
@@ -1933,6 +1938,10 @@ static inline void cpts_latch_pps_start(struct cpts *cpts)
 	v = READ_TCLR(cpts->odt2);
 	v |= BIT(11);
 	WRITE_TCLR(cpts->odt2, v);
+
+	/* disable tmr irq - overflow/hwts will do from now */
+	writel_relaxed(0, cpts->odt2->irq_ena);
+	__omap_dm_timer_write(cpts->odt2, OMAP_TIMER_WAKEUP_EN_REG, 0, 0);
 }
 
 static void cpts_latch_proc_init(struct cpts *cpts)
-- 
2.17.1

